### ISSUE TITLE
Complex Feature Criteria (v3.0)

### ISSUE THREAD

**Issue by FredKSchott** [FredKSchott/fflip#11]
A few months ago in #8 @suprememoocow shared how Gitter had extended fflip with a more expressive criteria system that supported "veto" voting (check out the PR for more info). I'm going to build off of that and outline a new syntax for defining feature criteria. My goal is to support "veto" voting (where a criteria not being met should force a disable) as well as some other requested features.

## Current Feature Format

```javascript
var ExampleFeaturesObject = {
  newFeatureRollout: {
    name: "A New Feature Rollout",  // user-defined, optional, for the user's documentation purposes only
    description: "Rollout of that new feature over the next month", // user-defined
    owner: "FredKSchott",  // user-defined
    criteria: {isPaidUser: false, percentageOfUsers: 0.50} // Match ALL criteria
  },
  // ...
}
```

## New Feature Criteria Format

```javascript
// Match ONE criteria
criteria: {isPaidUser: true} 
// Match ALL criteria
criteria: {isPaidUser: true, percentageOfUsers: 0.50} 
// Match ANY criteria
criteria: [{isPaidUser: true}, {isModernBrowsers: true}, {percentageOfUsers: 0.50}] 
// Match ANY criteria set (first criteria OR ALL criteria in the second criteria set)
criteria: [{isPaidUser: true}, {isModernBrowsers: true, percentageOfUsers: 0.50}] 
// Match ANY criteria, unless `isModernBrowsers` = false
criteria: [{isPaidUser: true}, {percentageOfUsers: 0.50}, {isModernBrowsers: true, $veto: true}] 
```

### Explained

- A **Criteria Set** is an object that describes a set of different criteria that must be met for a feature to be enabled. If ALL of those criteria are met, we say that the criteria set evaluates to true. Otherwise, the set evaluates to false. `{isPaidUser: true, percentageOfUsers: 0.50}`
- A **Criteria Collection** is an array of criteria sets and criteria collections. If ANY of its members evaluate to true, the entire collection evaluates to true. Otherwise, the collection evaluates to false. `[{isPaidUser: true}, {percentageOfUsers: 0.50}] `
- The `$veto` property exist on a member of a criteria collection. If that collection set evaluates to false, it will override the entire criteria collection to evaluate to false no matter what ANY of its other members evaluate to.
- A feature's `criteria` property can be either a criteria set or a criteria collection.

### Reasoning

This system has a ton of benefits over the current system, and a few over the veto system outlined in #8:

- Adds new OR logic that can match any criteria, instead of all criteria
- OR and AND logic can be combined to create really expressive criteria rules
- Features get to decide when a criteria is or isn't veto-worthy, so that they can be used in either scenario

Logically, the veto property just changes its criteria set to AND with everything in it's collection (instead of the default OR). So the name is a bit misleading, but it does a good job of communicating it in a relatable way. 

The other direction we could go would be to create a more expressive syntax that could actually support the mixing and matching of different logical operators "NOT", "OR NOT", "XOR", etc. But I'm not sure that use case exists or if the extra complexity on our end is worth it, at least in this first step.

---

Any and all feedback appreciated!

**Comment by suprememoocow** [comment]
Hi @FredKSchott, sorry I've been so scarce in the discussion around veto voting in #8. My current veto implementation is here: https://github.com/suprememoocow/fflip/tree/veto but it was done in a hurry without the forethought you've put into this. I really like the flexibility your design allows. 

I agree that using `NOT`/`AND`/`OR` operators etc is probably overkill.

**Comment by FredKSchott** [comment]
Working in this branch: https://github.com/FredKSchott/fflip/tree/v3-new-syntax

Currently only has support for new AND/OR logic, veto logic still to come.

**Comment by FredKSchott** [comment]
Merged via #13
